File: python/simple-cipher/simple_cipher.py
import random
import string

class Cipher(object):
    """Class to encode/decode text (just letters a-z) using a 
    modified version of the caesar cipher but with a key where
    the code can change for each letter in the encode/decode string.
    """
    def __init__(self, key=None):
        """Constructor, saves or sets up the cipher key
        Parameters:
        * key: What to use as the cipher key. If not specified will
        generate a random key.
        """
        if key == None:
            self.key = self.randomKey()
        else:
            self.key = key

    def encode(self, text):
        """Encode a message based on the key that was given/generated
        when the object was created.
        Parameters:
        * text: The message to encode.
        Returns:
        * Encoded version of the input string (text).
        """
        # Calculate the index into the lower case letters for both the 
        # current character of the string and the current character of the
        # key (can wrap around if text is longer than than the key). Add them
        # together (being careful to wrap around the other size of the alphabet
        # if we go past z). Then turn it back into a letter. Repeat for all 
        # characters in a a string and glue them together and return them.
        return ''.join([chr(ord('a') + ((ord(ch) - ord('a')) + (ord(self.key[i % len(self.key)]) - ord('a'))) % len(string.ascii_lowercase)) for i, ch in enumerate(text.lower())])

    def decode(self, text):
        """Decode a message based on the key that was given/generated
        when the object was created.
        Parameters:
        * text: The message to decode.
        Returns:
        * Decoded version of the input string (text).
        """
        # Calculate the index into the lower case letters for both the 
        # current character of the string and the current character of the
        # key (can wrap around if text is longer than than the key). Subtract
        # the key from the text. Then turn it back into a letter. 
        # Repeat for all characters in a a string and glue them together and 
        # return them.
        return ''.join([chr(ord('a') + ((ord(ch) - ord('a')) - (ord(self.key[i % len(self.key)]) - ord('a'))) % len(string.ascii_lowercase)) for i, ch in enumerate(text.lower())])
    
    def randomKey(self, length=100):
        """generate a key by randomly choosing character from the lower case
        letters (a-z) until you get to the specified length.
        Parameters:
        * length: Length of the desired string, 100 if not specified
        Returns:
        * Random letter string of the length provided.
        """
        # use random.choice to grab a random lowercase letter, fill a list
        # with the desired count of random letters then convert it into a string
        # and return the final value.
        return ''.join([random.choice(string.ascii_lowercase) for i in range(length + 1)])

File: csharp/simple-cipher/SimpleCipher.cs
using System.Text;

/// <summary>
/// Substitution Cipher
/// NOTE: Only characters a-z are encoded all others are passed through unchanged.
/// </summary>
public class SimpleCipher
{
    const int ALPHABET_LEN = (int)'z' - (int)'a' + 1; // a-z range
    string cipherKey; // encoding key.

    /// <summary>
    /// Class intializer, uses a random key
    /// </summary>
    public SimpleCipher()
    {
        cipherKey = RandomKey();
    }

    /// <summary>
    /// Class initializer, initialize with the given key
    /// </summary>
    /// <param name="key">The cipher key to use</param>
    public SimpleCipher(string key)
    {
        cipherKey = key;
    }

    /// <summary>
    /// Return the class' cipher key.
    /// </summary>
    public string Key
    {
        get
        {
            return cipherKey;
        }
    }

    /// <summary>
    /// Generate a random key when one is not provided. Change KEY_LENGTH to get larger or smaller keys.
    /// </summary>
    /// <returns>A string of random characters 'a' to 'z' of length KEY_LENGTH.</returns>
    private static string RandomKey()
    {
        const int KEY_LENGTH = 100;
        Random rnd = new();
        StringBuilder sb = new();
        for (int i = 0; i < KEY_LENGTH; i++)
        {
            sb.Append((char)(rnd.Next(ALPHABET_LEN) + (int)'a'));
        }
        return sb.ToString();
    }

    /// <summary>
    /// Encode the given text with the class' key
    /// </summary>
    /// <param name="plaintext">The text to encode</param>
    /// <returns>The encoded text</returns>
    public string Encode(string plaintext)
    {

        StringBuilder sb = new();
        for (int i = 0; i < plaintext.Length; i++)
        {
            if (plaintext[i] >= 'a' && plaintext[i] <= 'z')
            {
                int ch = (int)plaintext[i] - (int)'a';
                ch = (ch + (int)cipherKey[i % cipherKey.Length] - (int)'a') % ALPHABET_LEN;
                sb.Append((char)(ch + (int)'a'));
            }
            else
            {
                sb.Append(plaintext[i]); // Append normal if not in range.
            }
        }
        return sb.ToString();
    }

    /// <summary>
    /// Decode a given cipher using the class' key
    /// </summary>
    /// <param name="ciphertext">The text to decode</param>
    /// <returns>The decoded text</returns>
    public string Decode(string ciphertext)
    {
        StringBuilder sb = new();
        for (int i = 0; i < ciphertext.Length; i++)
        {
            if (ciphertext[i] >= 'a' && ciphertext[i] <= 'z')
            {
                int ch = (int)ciphertext[i] - (int)'a';
                // don't want to go less than zero so add in ALPHABET_LEN to be modded of but keep the value >= 0.
                ch = (ch + ALPHABET_LEN - ((int)cipherKey[i % cipherKey.Length] - (int)'a')) % ALPHABET_LEN;
                sb.Append((char)(ch + (int)'a'));
            }
            else
            {
                sb.Append(ciphertext[i]); // Append normal if not in range.
            }
        }
        return sb.ToString();
    }
}